### PR TITLE
Optimize TPL interface to rocsparse spmv by caching variables for repeated use

### PR DESCRIPTION
**Pull request by jczhang07**
Caching the temp variables is key to performance of repeated SpMV. I tested on one Crusher node with petsc microbenchmark (src/mat/tests/ex1k).

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-57f63c0c-7fff-468e-47c0-c6637abb2a4e">

 Average SpMV time (us) with matrix HV15R (of size 2M x 2M).
  | 1 rank | 8 ranks
-- | -- | --
TPL=OFF | 3743 | 636
TPL=ON | 14948 | 3713
TPL=ON + this PR | 2686 | 571

</b>

@lucbv  I hope to make `A.roc_spmv_helper` a private member, but I don't know how to access it in `spmv_rocsparse()` then.

### CODE REVIEW

**Comment by srajama1** [comment]
@jczhang07 Is this for several calls to SpMV? Is the caching basically avoiding some expensive allocations / initializations?

**Comment by jczhang07** [comment]
Yes, for repeated calls to SpMV as commonly seen in Krylov methods.

**Comment by lucbv** [comment]
This makes sense although I would like to think about all the TPLs together and try to come up with an implementation that takes care of all of this for cusparse, rocsparse, onemkl, etc...

**Comment by jczhang07** [comment]
That sounds reasonable.  But each TPL has its own APIs. In the end, we have to have separate implementations.

I suggest we could add standalone TPLs as we go, then we might come up with better ideas to consolidate them.

**Comment by lucbv** [comment]
Correct, I think handling the TPLs can be done in two parts

1. augment the CrsMatrix to hold additional data generated by the TPL
2. keep the individual kernel implementations for each TPL

**Comment by kokkos-devops-admin** [comment]
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

**Comment by kokkos-devops-admin** [comment]
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

**Comment by jczhang07** [comment]
@lucbv Testing pipeline is clean. Any comment on merging it?

**Comment by seanofthemillers** [comment]
@lucbv @jczhang07 Hi guys, is there an update on this? I have some overlapping changes that I would like to put in to help support rocsparse in Tpetra, but it would be nice to use this as a starting point (my existing work ignores cusparse).

**Comment by jczhang07** [comment]
> @lucbv @jczhang07 Hi guys, is there an update on this? I have some overlapping changes that I would like to put in to help support rocsparse in Tpetra, but it would be nice to use this as a starting point (my existing work ignores cusparse).

@seanofthemillers On my end, I just need reviewers and approvals to get it merged :)

**Comment by seanofthemillers** [comment]
@lucbv I had problems in my implementation with adding the raw descriptor to the CrsMatrix. In an application like Tpetra, these KokkosSparse::CrsMatrix objects are generated on the fly (i.e. not stored and re-used), so we would have an issue where the KokkosSparse::CrsMatrix would destroy the descriptor on destruction. The Tpetra design implies that the Tpetra::CrsMatrix should own the descriptor, but it would be nice to automate the generation of the descriptors via kokkos-kernels to keep raw rocsparse/cusparse code out of Tpetra. The shared_ptr method used in this PR gives us a nice workaround.

**Comment by lucbv** [comment]
@seanofthemillers the shared pointer can be stored in the handle or controls of Kokkos Kernels then, although I have to say having Tpetra destroy and create the matrix on the fly is a but annoying even if I understand the idea.
We are just not keen on creating a new data structure when we already have two for similar purpose in the library.

I think Tpetra might be destroying our controls and handles after performing the local kernel but that's really a problem for them to figure out.

**Comment by kokkos-devops-admin** [comment]
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

**Comment by jczhang07** [comment]
@lucbv  I forgot this MR hadn't been merged yet until I tested a SpMV benchmark on Frontier recently and saw the terrible performance again (https://github.com/kokkos/kokkos-kernels/pull/1469#issue-1306297831).  

I rebased the MR off KK/develop.  Could you have a look and get it merged?

File: sparse/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp [hunk 1]
```diff
@@ -404,76 +487,85 @@ void spmv_rocsparse(const Kokkos::HIP& exec,
...
-  KOKKOS_IMPL_HIP_SAFE_CALL(hipFree(tmp_buffer));
 
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_destroy_dnvec_descr(vecY));
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_destroy_dnvec_descr(vecX));
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_destroy_spmat_descr(Aspmat));
-  KOKKOS_ROCSPARSE_SAFE_CALL_IMPL(rocsparse_destroy_mat_descr(Amat));
+    KOKKOS_IMPL_HIP_SAFE_CALL(
+        hipMalloc(&spmv_helper->buffer[idx], spmv_helper->buffer_size[idx]));
+
+#if KOKKOSSPARSE_IMPL_ROCM_VERSION >= 50400
```
Comment (jczhang07) [inline sparse/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp:545]: @lucbv Note that I removed the old ` KOKKOSSPARSE_IMPL_ROCM_VERSION >= 60000` stuff and prefer `rocsparse_spmv_ex` over `rocsparse_spmv`

**Comment by jczhang07** [comment]
@brian-kelley Could you test this PR in CI?

**Comment by lucbv** [comment]
@jczhang07 this work was eventually done via PR #2126 I will close this one but if you notice issues with how we use the rocSparse TPL please let us now

**Pull request closed without merging**